doc/adr/0004-pab-indexing-solution-integration.rst

```diff
@@ -0,0 +1,183 @@
+                 >>= submitUnbalancedTx . Constraints.adjustUnbalancedTx
+
+      -- Step 2
+      _ <- awaitTxConfirmed (getCardanoTxId ledgerTx)
+
+      -- Step 3
+      let refs = Map.keys
+                 $ Map.filter ((==) address . txOutAddress)
+                 $ getCardanoTxUnspentOutputsTx ledgerTx
+      case refs of
+          []                   -> throwing _ScriptOutputMissing pk
+          [outRef] -> do
+              -- Step 4
+              ciTxOut <- unspentTxOutFromRef outRef
+              pure (outRef, ciTxOut, inst)
+          _                    -> throwing _MultipleScriptOutputs pk
+
+
+Here's an outline of the contract's steps:
+
+1. Creates a transaction and submits it to the node
+2. Waits for transaction to be confirmed
+3. Finds the first UTXO of that transaction (return type `TxOutRef`)
+4. Queries the plutus-chain-index to get the `ChainIndexTxOut` out of that `TxOutRef`
+
+The problem is that the `ciTxOut` variable in step 4 will almost always result in `Nothing`.
+
+Why? Here’s some context.
+
+The PAB listens to the local node and stores blockchain information in memory such as the status of transactions, the status of transaction outputs, the last synced slot, the current slot, etc., in a variable of type `BlockchainEnv`.
+The `awaitTxConfirmed` is actually querying the state of `BlockchainEnv` and waits until the status of the transaction transitions to `Confirmed`.
+
+Meanwhile, plutus-chain-index (our main indexing component at the time of this writing) is also listening to incoming blocks from the local node and indexes them into a database.
+The indexed data can be queried using the REST API interface.
+
+This brings up the main issue: the PAB and plutus-chain-index each listen to the same source of information (a local Cardano node), but each index the information at different speeds.
+For a dApp developer writing off-chain code using the Contract API, there is no abstraction for handling multiple sources of truth.
+
+Currently, in the best case scenario (fully synced PAB and plutus-chain-index), plutus-chain-index will always trail behind the in-memory storage of the PAB by a few seconds.
+Therefore, even in this scenario, querying the plutus-chain-index with `unspentTxOutFromRef` in the above contract has a high probability of returning `Nothing`.
+
+Decisions
+---------
+
+The best solution is probably a combination of the `Alternative solutions`_ described below.
+However, we will mainly choose the `Query functions should interact with a single source of truth`_ solution.
+
+* We will replace `plutus-chain-index` with `Marconi` as PAB's indexing component
+
+* We will move out the Blockchain information indexed by PAB in Marconi
+
+* We will add new indexers in Marconi in order to replicate the information indexed by `plutus-chain-index`
+
+* We will adapt the architecture of Marconi (which will become our new indexing component) to support waiting queries
```

Comment:
Currently, we have only implemented indexers which stores information in the database. But the indexer also have a notification field, right? Then this notification field allows an indexer to support waiting queries (like `awaitTxConfirmed`).

I used the word `adapt` because we haven't officially made a decision on how the notification system will work. How would you rephrase it?

Reply:
The phrasing is fine, I was just curious what you meant. You are making it sound a bit like a teaser of things to come in the next episode. Which is fine by me. :)